### PR TITLE
Basic JSDoc types - Plot.plot and all Marks

### PR DESCRIPTION
**Pull request by duaneatat**
This PR uses a `types.ts` file for authoring the types (this is just for convenience, since I find it easier to write typescript types than use the JSDoc syntax), and we import those types as jsdoc tags where appropriate.

This PR covers Plot.plot and all of the marks.

Known issues:
- Some types are too broad, e.g. tickX doesn't accept x1 and y1, but those appear in the completions

TODO:
- [ ] Plot.image : move the description of options from the `### Image` section to the Plot.image jsdocs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1245,6 +1245,8 @@ The given *options* are passed through to these underlying marks, with the excep
 
 #### Plot.boxX(*data*, *options*)
 
+
+
```
Comment (duaneatat) [inline README.md:1249]: I plan to trim these new lines in the jsdoc-to-readme script

File: test/plots/index.html [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <!DOCTYPE html>
 <head>
-  <meta charset="utf-8">
```
Comment (duaneatat) [inline test/plots/index.html:3]: Odd, not sure why this change was necessary. Will take a look tomorrow

Comment (Fil) [reply test/plots/index.html:3]: I've had the same experience on a different branch. Could it be that the github action has changed?

Comment (Fil) [reply test/plots/index.html:3]: put this tiny change in a separate PR #1059

Comment (duaneatat) [reply test/plots/index.html:3]: thank you! I'll merge main in once that's in

**Comment by Fil** [comment]
I think I'd like to have all the types imports at the top of each file, as I was trying to do with the "import types" in #1005.


Maybe for later: I've tried to add a BarYOptions to the `barY(data, options)` mark, by creating
```
export interface BarYOptions extends MarkOptions {
  offset: null | "expand" | "center" | "wiggle" | function…;
  order?: "sum" | "value" | …;
}
```

Of course, it would have to be extended to cover all the stacking options, which means I'd want to write something like `BarYOptions = MarkOptions & StackYOptions;`. Not sure if this can be done with "type" or "interface", though.

**Comment by Fil** [comment]
I've just noticed that Plot.window(X/Y) and Plot.normalize(X/Y) have no jsdocs at all.

**Review by Fil — APPROVED** [review]
lgtm!

File: src/marks/area.js [hunk 3]
```diff
@@ -85,8 +92,10 @@ export class Area extends Mark {
  * [Plot.areaX](https://github.com/observablehq/plot/blob/main/README.md#plotareaxdata-options)
  * is used in the vertical orientation where the baseline and topline share *y*
  * values.
+ * @param {Data} data
+ * @param {Options} options
  */
-export function area(data, options) {
+export function area(data, options = undefined) {
```
Comment (duaneatat) [inline src/marks/area.js:98]: In order to have editors know that options is optional, we had to add these default values for options throughout. Without this, the typescript language server will consider options required:

image


Adding ? to the @param also didn't seem to help:

image

Only adding a default value will produce the ? in the signature help:
image

Comment (duaneatat) [reply src/marks/area.js]: @mbostock pointed out that this can be done by using the proper JSDoc syntax for optional parameters (`@param {Options=} options`) 🙏

**Comment by mbostock** [comment]
Probably superseded by #1320 which goes the d.ts route. Close?

**Comment by Fil** [comment]
I'm good with closing, but let's keep the branch around for a while. There are a few things that I'll want to refer to, for example the to replace the D3 type:

```
/**
 * A symbol object with a draw function
 * @link https://github.com/d3/d3-shape/blob/main/README.md#custom-symbol-types
 */
type SymbolObject = {draw: (context: CanvasPath, size: number) => void};
```

**Comment by mbostock** [comment]
Superseded by #1343.
